### DISCUSSION TITLE
inconsistencies in unwrapped interferogram between tiles

### DISCUSSION THREAD

**Discussion by RShamshiri** [opera-adt/tophu#14] [Q&A]
Dear All,

I am testing Tophu as the original phase unwrapping results in several phase jumps. To do so, I am running the following command from the branch located at "https://github.com/gmgunter/tophu/tree/conncomp-labels":

`uw, cc = tophu.multiscale_unwrap(int, cor, 5*28, unwrap=tophu.unwrap.SnaphuUnwrap(), downsample_factor=(15, 15), ntiles=(3,3), tile_overlap=[700, 700])`

and here are the results: unwrapped interferogram and the connected component:
![uw_ds1515_highres_ovl700_t33](https://github.com/opera-adt/tophu/assets/105787277/e782db0f-2d6d-4e2d-a4e0-80d65495fca2)
![cc_ds1515_highres_ovl700_t33](https://github.com/opera-adt/tophu/assets/105787277/dd9ae1bc-9011-41ce-8259-7f9c9b689394)

I ran `multiscale_unwrap` with various down-sampling factors and tiles numbers, but the unwrapped phase has inconsistencies between the tiles. I wonder if I ran it correctly or need to adjust the settings for a better result. 

Here are the unwrapped interferogram and the connected component map obtained using the original phase unwrapping in Snaphu:
![221221-230104_ionoCorrected_unwrapped_int_isce2](https://github.com/opera-adt/tophu/assets/105787277/8ef78313-ac5d-4a06-971a-5252cfe1c139)
![221221-230104_conncomp_isce2](https://github.com/opera-adt/tophu/assets/105787277/68c7f9f3-a75a-494f-b281-18b6f3d712a6)

and here is the coherence map:
![221221-230104_cor_isce2](https://github.com/opera-adt/tophu/assets/105787277/52825751-54f5-440f-92d7-c1a68b03108c)

I would also like to know if the multiscale unwrapping method works on the interferograms with such poor coherence in general.  

Kind regards
Roghayeh

**Comment by gmgunter** [comment]
Hi @RShamshiri,

Any pixels in the unwrapped phase whose connected component label is zero are typically assumed to be unreliable. tophu doesn't currently try to apply any correction to those pixels.

I think that's why you're seeing those large phase jumps in the tophu result. The discontinuities appear to be occurring at the boundaries of connected components or along tile boundaries within the "invalid" regions (i.e. regions where the connected component labels were all zero).

I usually mask out these "invalid" pixels when visualizing the unwrapped output in order to just look at the pixels that the unwrapper deemed reliable. You might find that the `multiscale_unwrap()` outputs look a bit more reasonable if you do something similar.

```python
import numpy as np
import matplotlib.pyplot as plt
import tophu

# Compute unwrapped phase and connected components
unw, cc = tophu.multiscale_unwrap(...)

# Treat conncomp=0 pixels as invalid
unw_masked = np.ma.array(unw, mask=(cc == 0))

# Plot the unwrapped phase with invalid values masked
plt.imshow(unw_masked)
```

It's concerning to me, though, that it looks like vanilla SNAPHU assigned a nonzero connected component label to a much larger portion of the image than tophu did. I wouldn't expect that -- under the hood, tophu is basically running SNAPHU independently on each tile. Generally, the overall area unwrapped by tophu has been similar to SNAPHU in my testing.

Did you apply some different configuration options when you ran SNAPHU?

For some context, tophu operates under the assumption that each individual connected component in each tile is unwrapped in a self-consistent manner, but that the relative phase between individual tiles (or even individual connected components within the same tile) may be incorrect. If this is true, then we should be able to apply a constant phase shift to each component in order to correct for the relative phase errors.

In order to estimate the correct phase shift to apply, tophu unwraps a filtered, downsampled replica of the original interferogram as a single tile and uses that as a reference.

In general, it's not safe to expect that applying a similar constant phase shift to pixels with connected component label = 0 will work, because the relative phase between pixels within this region is likely to be incorrect. That's why `multiscale_unwrap()` doesn't try to apply any correction to such pixels. Of course, this does make tophu somewhat reliant on the underlying unwrapper to produce useful connected components, which is not always the case in practice.

Hope that helps.

**Comment by RShamshiri** [comment]
Hi @gmgunter!

Thank you for your feedback!

As you suggested, I removed the pixels labeled zero, and here are the masked unwrapped interferograms from Tophu, the original Snaphu, and their differential map:

![image](https://github.com/opera-adt/tophu/assets/105787277/5a6911cc-cda5-4d36-b365-f0d7d943c64e)
![image](https://github.com/opera-adt/tophu/assets/105787277/26123453-2276-47c4-ab4c-16c85e63c8e7)
![image](https://github.com/opera-adt/tophu/assets/105787277/29289add-256d-49b1-8e26-b9424bec335d)

Here, I plotted the pixel values along the dashed line shown in the first two maps:  
![image](https://github.com/opera-adt/tophu/assets/105787277/b64f5042-4d06-43bd-9ea2-60405a1a9724)

as shown in the maps, the results are almost identical, and the jumps are still there! I tried applying Tophu with different overlapping (700 and 0) and down-sampling (3/7/15) factors, but I did not see any significant improvement! I wonder if increasing the down-sampling factor would help! as this would make the reference plane (the coarse resolution interferogram) smoother, thereby reducing the potential jumps in it.

I note that I have not changed the configuration options while running the original Snaphu, but applied spatial filtering with a larger window size before phase unwrapping in ISCE2 due to poor coherence. 

Kind regards
Roghayeh

**Reply by scottstanie** [reply]
thanks for the detailed report! 
Would you mind posting the original interferogram and the coherence map that you're using?

**Reply by RShamshiri** [reply]
Thanks for your response. You can access the interferogram (before and after applying spatial filtering) and the coherence files by following this link:

https://www.dropbox.com/scl/fi/u1gf2gxajpr1vwkc8ghdc/ifgcor.zip?rlkey=zg6mdm3b70ovj5eqci8d7ma4j&dl=0

**Comment by RShamshiri** [comment]
Hi @scottstanie 

I'm following up on the files you requested to confirm if they were received in good order. Please let me know if there were any issues with the files.

Thank you
